coreos-koji-tagger/coreos_koji_tagger.py

```diff
@@ -264,69 +288,111 @@ def __call__(self, message: fedora_messaging.api.Message):
-        # to the tag before we can add the specific build to the tag
-        pkgstoadd = []
-        for build in buildstotag:
-
-            # Find the some defining information for this build.
-            buildinfo = get_rich_info_for_rpm_string(build, arch=False)
-
-            # Check to see if the koji pkg is already covered by the tag
-            if buildinfo.name not in pkgsintag:
-                pkgstoadd.append(buildinfo.name)
+        # compute the pkgstoadd and the buildstotag
+        pkgstoadd   = desiredpkgs - pkgsintag
+        buildstotag = desiredbuildids - currentbuildids 
 
         # Log if there is nothing to do
         if not pkgstoadd and not buildstotag:
             logger.info(f'No new builds to tag.. going back to sleep')
             return
 
-        # Add the needed packages to the tag if we have credentials
+        # Make sure all packages desired are in the pkglist
         if pkgstoadd:
             logger.info('Adding packages to the '
                         f'{self.target_tag} tag: {pkgstoadd}')
             if self.keytab_file:
-                add_pkgs_to_tag(tag=self.target_tag,
-                                pkgs=pkgstoadd,
-                                owner=self.koji_user)
+                with self.koji_client.multicall(strict=True) as m:
+                    for pkg in pkgstoadd:
+                        m.packageListAdd(self.target_tag, pkg,
+                                         owner=self.koji_user)
                 logger.info('Package adding done')
 
         # Perform the tagging for each release into the intermediate
         # tag for that release if we have credentials
         if buildstotag:
-            releasevers = set(buildsinfo.values())
-            for releasever in releasevers:
-                tag = self.intermediate_tag.format(releasever=releasever)
-                buildstotagforthisrelease = \
-                    [x for x in buildstotag if buildsinfo[x] == releasever]
-                if buildstotagforthisrelease:
-                    logger.info('Tagging builds into the '
-                                f'{tag} tag: {buildstotagforthisrelease}')
-                    if self.keytab_file:
-                        tag_builds(tag=tag, builds=buildstotagforthisrelease)
+            # Get a set of tuples of build name to tag to tag into
+            tuples = [(self.intermediate_tag.format(
+                        releasever=buildsinfo[x].releasever),
+                        buildsinfo[x].nvr)
+                        for x in buildstotag]
+            logger.info('Tagging the following (tag, nvr): \n\t%s' %
+                                            '\n\t'.join(map(str, tuples)))
+            if self.keytab_file:
+                with self.koji_client.multicall(strict=True) as m:
+                    for (tag, nvr) in tuples:
+                        m.tagBuild(tag=tag, build=nvr)
             logger.info('Tagging done')
```

Comment:
This line feels a bit disingenuous; should it perhaps be moved inside of the `if self.keytab_file` block so that it only prints if an action was taken?

Reply:
indented!